Resolve #25 to lint project according to flake8

wearymanateevedaknotstabooresisting opened the pull request:
Resolves #25

# Changes

Please refer to the topics addressed in #25. The PR aims to lint according to flake8, replacing star-imports and trying to wrap around maximum line length of 79.

# Implementation

- PR has been rebased to `mypy` branch which introduces `setup.cfg`
- flake8 configuration file has been added to `setup.cfg`. Currently only ignoring Python virtual environment wildcard directory. Can be extended, later on - may be to ignore wanted imports like the logging configuration. I'd opt regarding re-writing this import and explicitly calling a function at the very beginning.
- E501 maximum line lengths have been refactored. This sometime lead to strange formatting with escape line breaks etc. Long documentation links had to be removed, I'd suggest to move them into a wiki or a documentation file.
- Please refer to `setup.cfg` flake8 configuration and its ignored warnings/formatting hints for details.

wearymanateevedaknotstabooresisting commented:
@provinzio and @shredEngineer: help wanted, if you have preferred coding / formatting guide lines.

provinzio commented:
@wearymanateevedaknotstabooresisting As you are more of a GitLab user like me, you might be interested in this info. I noticed earlier that GitHub calls PR which are WIP draft. You can create a PR as draft like shown [here](https://github.blog/2019-02-14-introducing-draft-pull-requests/) or change the status in a PR on the right side (see image below)

![grafik](https://user-images.githubusercontent.com/20270870/110371050-bd7eb680-804c-11eb-94ab-5fb7c7be2b0e.png)

wearymanateevedaknotstabooresisting commented:
> As you are more of a GitLab user like me

True, in GitLab you can Prefix the PR Name with `WIP ` or also use such a button. Did not see it in my GitHub dark mode when opening 👍

wearymanateevedaknotstabooresisting commented:
I rebased to `mypy` branch and fixed the `flake8` import formatting hints to module imports with explicit object calls. Please review and maybe have a look into the `mypy src/main.py` hints which I couldn't wrap my head around, yet.

provinzio commented on requirements.txt:
```diff
@@ -1,7 +1,16 @@
+## The following requirements were added by pip freeze:
```
should we split the file into `req...` and `req..._dev`? I don't know whats the best practice here.

provinzio replied:
I splitted the files (220decac1c7c3f11eb88c8c404ab095ec8cbd8ee)

wearymanateevedaknotstabooresisting replied:
Actually, best practice is having a `requirements.txt` for dev and a `setup.py` for production. However, I like simple pip requirements so much more myself. So, I am fine with having two files. Just keep in mind to keep them up-to-date, both.

provinzio reviewed (commented):
Good work so far. I am unfortunately short on time at the moment. My personal todo list:
- [ ] have a look at the flake8 documentation and decide which options we want to set
- [x] have a look at the left over warnings

provinzio commented:
I added `isort` for linting and created a workflow which runs all our linters on push. I am not completly satisfied with the current workflow
- [ ] mypy is called quite hacky. Just running `mypy . ` results in an `re.error: nothing to repeat at position 0`. I don't know how to resolve this. Any ideas?
- [ ] We might be able to increase the speed with a dockerfile, but I don't know how github handles the workflows internally... so this might not be worth is effort.

wearymanateevedaknotstabooresisting commented:
> * [ ]  mypy is called quite hacky. Just running `mypy . ` results in an `re.error: nothing to repeat at position 0`. I don't know how to resolve this. Any ideas?

I had to call `mypy src/main.py` on my machine. First time for me, using mypy in a project, though.

The workflows are docker based? I could create a `Dockerfile` but I am not understanding how this is linked to `mypy`.

provinzio commented:
On my machine `mypy src/main.py` only runs the linter for the file `main.py`.

It looks like the github workflows can be setup with dockerfiles (like in GitLab). Docker has no direct connection with mypy.

provinzio commented:
@wearymanateevedaknotstabooresisting Oh my dear. It took me way to long to figure that out. The `mypy` exclude config option is supposed to be a regex https://github.com/provinzio/CoinTaxman/pull/27/commits/56f8a94884a8144bdc3779a0878289c2e31ff133. This caused the above error.

So as far as I am considered, we are more than done here, implementing not just `flake8` but also `isort` and `mypy`. I'll give it a last look and merge afterwards.

wearymanateevedaknotstabooresisting commented:
Great work finalizing this PR! Thank you very much :-)
